x/mongo/driver/topology/server.go

```diff
@@ -292,7 +293,29 @@ func (s *Server) ProcessHandshakeError(err, ctxErr error, startingGenerationNumb
 		return
 	}
 
-	isTimeoutOrCanceled := func(err error) bool {
+	isCtxTimeoutOrCanceled := func(ctx context.Context) bool {
+		if ctx == nil {
+			return false
+		}
+
+		if ctx.Err() != nil {
+			return true
+		}
+
+		// In some networking functions, the deadline from the context is used to determine timeouts
```

Comment:
Thank you for the helpful explanation! That is interesting, and sounds like it could be possible buggy behavior in the net package if it inconsistently sets context errors.